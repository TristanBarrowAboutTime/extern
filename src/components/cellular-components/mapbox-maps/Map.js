import * as React from 'react';
import {useState, useRef} from 'react';
import MapGL, {Source, Layer} from 'react-map-gl';
import styled from 'styled-components';

const MAPBOX_TOKEN = 'pk.eyJ1IjoidHJpc3Rhbi1hYm91dHRpbWUtZGV2IiwiYSI6ImNramE4djMzczQza20ycW0wenF2MXlpN2wifQ.Yl6NjsxYBTyWGsptG40luw';

const MapContainer = styled.div`
    width: 100%;
    height: 100%;
`;

const Map = (props) => {
    const binding = useMap();

    return (
        <MapContainer>
            <MapGL
                {...props.viewport}
                width="100%"
                height="100%"
                mapStyle="mapbox://styles/tristan-abouttime-dev/ckjac5z076e4q1al74uthbd1t"
                onViewportChange={props.changeViewPort}
                mapboxApiAccessToken={MAPBOX_TOKEN}
                onClick={binding.onClick}
                ref={binding.mapRef}
            >
                <Source 
                    id='worksite-data' 
                    type='geojson' 
                    data={props.worksiteData}
                >
                    <Layer 
                        id='worksites'
                        source='worksite-data'
                        type='circle'
                        paint={{
                            'circle-radius': 10,
                            'circle-color': '#D86833'
                        }}
                    />
                </Source>
                <Source 
                    id='employee-data' 
                    type='geojson' 
                    data={props.employeeData}
                >
                    <Layer 
                        id='employees'
                        source='employee-data'
                        type='circle'
                        paint={{
                            'circle-radius': 10,
                            'circle-color': '#000000'
                        }}
                    />
                </Source>
            </MapGL>
        </MapContainer>
    );
}

/**
 * export const clusterLayer = {
  id: 'clusters',
  type: 'circle',
  source: 'earthquakes',
  filter: ['has', 'point_count'],
  paint: {
    'circle-color': ['step', ['get', 'point_count'], '#51bbd6', 100, '#f1f075', 750, '#f28cb1'],
    'circle-radius': ['step', ['get', 'point_count'], 20, 100, 30, 750, 40]
  }
};
 */

const useMap = () => {

    const mapRef = useRef(null);

    const onClick = (event) => {};
    
    return {
        mapRef,
        onClick
    }
}

export default Map;