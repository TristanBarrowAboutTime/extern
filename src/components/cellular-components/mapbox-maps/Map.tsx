import * as React from 'react';
import { useRef } from 'react';
import MapGL, { FlyToInterpolator, Layer, LayerProps, Source, SourceProps } from 'react-map-gl';
import styled from 'styled-components';
import MapButtons from './MapButtons';

export enum BaseMap { 
    BASIC = 'mapbox://styles/mapbox/streets-v11',
    // SATELLITE = 'mapbox://styles/mapbox/satellite-v9',
    SATELLITE_STREET = 'mapbox://styles/mapbox/satellite-streets-v11',
}

const Container = styled.div`
    position: relative;
    width: 100%;
    height: 100%;
`;

const MapControlls = styled.div`
    position: absolute;
    top: 0;
    left: 0;


`;


export type Viewport = {
    latitude: number
    longitude: number
    zoom: number
    bearing: number
    pitch: number
    transitionDuration: number 
    onViewportChange: (viewport: any) => void
    ref: React.Ref<any>
    height: string
    width: string
    mapStyle: BaseMap
    transitionEasing: (x: number) => number
    
}

export type MapActions = {
    setCords: (lat: number, long: number) => void
    setZoom: (zoom: number) => void
    incZoom: () => void
    decZoom: () => void
    setToStreet: () => void
    setToSatellite: () => void
}

export type MapConfiguration = {
    source: SourceProps
    layers: LayerProps[]
}

type MapProps = {
    viewport: Viewport
    actions: MapActions
    mapConfigs?: MapConfiguration[] 
}

const Map = (props: MapProps) => {
    const { viewport, actions, mapConfigs = [] } = props;
    return (
        <Container>
            <MapGL
                {...viewport}
                mapStyle={viewport.mapStyle.toString()}
                mapboxApiAccessToken={process.env.REACT_APP_MAP_TOKEN}
            >
                {mapConfigs.map((config: MapConfiguration) => {
                    return <Source {...config.source}>
                        {config.layers.map((layer: LayerProps) => {
                            return <Layer {...layer}/>
                        })}
                    </Source>
                })}
            </MapGL>
            <MapButtons
                actions={actions}
                selectedMapStyle={viewport.mapStyle}
            />
        </Container>
    );
}

type UseWithMapArgs = {
    initialBaseMap?: BaseMap
    maxZoom?: number
    minZoom?: number
}


export const useWithMap = ({
    initialBaseMap = BaseMap.BASIC,
    maxZoom = 22,
    minZoom = 1
}: UseWithMapArgs) => {
    const ref = useRef(null);
    const [viewport, setViewportTo] = React.useState({
        latitude: 39.98819048878927,
        longitude: -111.75807962919087,
        zoom: 12,
        bearing: 0,
        pitch: 0,
        transitionDuration: 500,
        transitionEasing: (x: number): number => (x < 0.5 ? 4 * x * x * x : 1 - Math.pow(-2 * x + 2, 3) / 2),

    });
    
    const [baesMap, setBaseMapTo] = React.useState(initialBaseMap);
    const setToSatellite = () => setBaseMapTo(BaseMap.SATELLITE_STREET);
    const setToStreet = () => setBaseMapTo(BaseMap.BASIC);

    const changeViewPort = (viewPort: any) => {
        setViewportTo({
            ...viewPort,
            pitch: 0,
            bearing: 0,
            transitionDuration: 500,
        });
    }


    const setCords = (lat: number, long: number) => {
        changeViewPort({
            ...viewport,
            latitude: lat,
            longitude: long
        });
    }

    const setZoom = (zoom: number) => {
        let newZoom = zoom;
        if (zoom <= minZoom) {
            newZoom = minZoom;
        }
        if (zoom >= maxZoom) {
            newZoom = maxZoom;
        }
        changeViewPort({
            ...viewport,
            zoom: newZoom
        });
    }

    const incZoom = () => setZoom(viewport.zoom <= maxZoom ? viewport.zoom + 1 : viewport.zoom);
    const decZoom = () => setZoom(viewport.zoom >= minZoom ? viewport.zoom - 1 : viewport.zoom);


    return {
        viewport: {
            ...viewport,
            onViewportChange: setViewportTo,
            ref,
            height: '100%',
            width: '100%',
            mapStyle: baesMap 
        },
        actions: {
            setCords,
            setZoom,
            incZoom,
            decZoom,
            setToStreet,
            setToSatellite,
        }

    }
}

export default Map;