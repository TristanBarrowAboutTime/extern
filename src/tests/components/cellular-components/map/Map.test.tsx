import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { renderHook, act } from '@testing-library/react-hooks';
import Map, { useWithMap, Viewport, MapActions, BaseMap } from '../../../../components/cellular-components/map/Map';

let viewport: Viewport | null = null;
let actions: MapActions | null = null;

beforeEach(() => {
    viewport = {
        latitude: 1,
        longitude: 55,
        zoom: 10,
        bearing: 0,
        pitch: 0,
        transitionDuration: 500, 
        ref: null,
        height: '100%',
        width: '100%',
        mapStyle: BaseMap.BASIC,
        onViewportChange: jest.fn(),
        transitionEasing: jest.fn()
    }

    actions = {
        setCords: jest.fn(),
        setZoom: jest.fn(),
        incZoom: jest.fn(),
        decZoom: jest.fn(),
        setToStreet: jest.fn(),
        setToSatellite: jest.fn(),
    }
});

afterEach(() => {
    viewport = null;
    actions = null;
});




describe('Map', () => {
    it('renders correctly', () => {
        const tree = renderer.create(
            <Map 
                viewport={viewport as Viewport}
                actions={actions as MapActions}
                mapConfigs={[]}
            />
        ).toJSON();   
        expect(tree).toMatchSnapshot();    
    });
});



describe('Use With Map', () => {
    it('exists', () => {
        const { result } = renderHook(() => useWithMap({}));
        expect(result.current.viewport.latitude).toBeTruthy(); 
        expect(result.current.viewport.longitude).toBeTruthy(); 
        expect(result.current.viewport.zoom).toBeGreaterThan(0); 
        expect(result.current.viewport.bearing).toBe(0); 
        expect(result.current.viewport.pitch).toBe(0); 
        expect(result.current.viewport.transitionDuration).not.toBeUndefined(); 
        expect(result.current.viewport.transitionEasing).toBeTruthy();
    });
    it('Increments Zoom', () => {
        const { result } = renderHook(() => useWithMap({
            minZoom: 2,
            maxZoom: 8,
            initialZoom: 5,
        }));
        const zoom = result.current.viewport.zoom;

        act(() => {
            result.current.actions.incZoom();
        });
        expect 
    })
});