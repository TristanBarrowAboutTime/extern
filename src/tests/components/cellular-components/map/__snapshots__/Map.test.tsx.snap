// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map renders correctly 1`] = `
<div
  className="sc-gKAblj bEUqUl"
>
  <div
    style={
      Object {
        "cursor": "grab",
        "height": "100%",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "height": "100%",
          "position": "relative",
          "width": "100%",
        }
      }
    >
      <div
        className=""
        style={
          Object {
            "height": "100%",
            "overflow": "hidden",
            "position": "absolute",
            "visibility": "inherit",
            "width": "100%",
          }
        }
      />
      <div
        style={
          Object {
            "height": 0,
            "overflow": "visible",
            "width": 0,
          }
        }
      />
    </div>
  </div>
  <div
    className="sc-bdnylx ctNIPE"
  >
    <div
      className="sc-dlnjPT jtZTbX"
      onClick={[MockFunction]}
    >
      Map
    </div>
    <div
      className="sc-gtssRu dnZDtQ"
      onClick={[MockFunction]}
    >
      Satellite
    </div>
  </div>
  <div
    className="sc-hKFyIo dcvdQf"
  >
    <div
      className="sc-eCApGN jeVmJR"
      onClick={[MockFunction]}
    >
      +
    </div>
    <div
      className="sc-jSFkmK iJtdXc"
      onClick={[MockFunction]}
    >
      -
    </div>
  </div>
</div>
`;
